File: Script/WebSocket/EventListenerList/SendMessage.php
<?php
require_once __DIR__ . "/../EventListener.php";
require_once __DIR__ . '/../../Util.php'; 
require_once __DIR__ . '/../WebSocket.php'; 
require_once __DIR__ . '/../WebSocketMysql.php'; 
require_once Util::CONFIG_PATH;

use Ratchet\ConnectionInterface;

class SendMessage implements EventListener{
    public function OnCall(ConnectionInterface $from, $json_data) {
        if(!isset($from->userId)){
            $json_data[STATUS] = STATUS_ERROR; 
            echo "userId not setup \n";
            $from->send(json_encode($json_data));
        } else if(!isset($from->channel) || $from->channel === NOT_SETUP){
            $json_data[STATUS] = STATUS_ERROR; 
            echo "user channel is not setup \n";
            $from->send(json_encode($json_data));
        } else {
            $json_data[STATUS] = STATUS_SUCCESS; 
            $json_data[KEY_DATETIME] = date("Y-m-d H:i:s");
            if(WebSocketMysql::Instance()->isDMChannel($from->channel)){
                $tableName = "dm_". $from->channel;
                WebSocketMysql::Instance()->addChatData($tableName, $from->userId, $json_data["message"]);
                foreach ( WebSocket::Instance()->channels[$from->channel] as $client){
                    if($from === $client) {
                        $json_data["isSelf"] = true;
                    } else {
                        $json_data["isSelf"] = false;
                    }

                    echo "channel: {$client->channel} userId: {$client->userId}" . json_encode($json_data) . "\n";
                    $client->send(json_encode($json_data));
                }
            }

        }
    }
}
?>

File: Script/WebSocket/SocketScript/java/event/listener/SendMessage.java
package event.listener;

import com.sun.org.apache.bcel.internal.generic.LOOKUPSWITCH;
import org.example.MysqlManager;
import org.example.SocketConnection;
import org.example.SocketEventListener;
import org.example.Util;
import org.json.JSONObject;

import java.nio.channels.SocketChannel;
import java.time.LocalDateTime;
import java.time.format.DateTimeFormatter;

public class SendMessage implements SocketEventListener.EventListener {
    @Override
    public void run(SocketChannel channel, JSONObject jsonObject) {
        if(!SocketConnection.Instance().userMap.containsKey(channel) ||
        SocketConnection.Instance().userMap.get(channel).getUserId() == SocketConnection.UserData.NOT_SETUP){
            System.out.println("userId not setup");
            jsonObject.put(SocketEventListener.eKey.STATUS.toString(), "error");
            SocketConnection.trySendMessage(channel, jsonObject.toString());
        } else if (SocketConnection.Instance().userMap.get(channel).getChannelId() == SocketConnection.UserData.NOT_SETUP) {
            System.out.println("channel not setup");
            jsonObject.put(SocketEventListener.eKey.STATUS.toString(), "error");
            SocketConnection.trySendMessage(channel, jsonObject.toString());
        } else {
            String datetime = Util.getCurrentDateTime();
            jsonObject.put(SocketEventListener.eKey.STATUS.toString(), "success");
            jsonObject.put(SocketEventListener.eKey.DATETIME.toString(), datetime);

            //현제 이부분 체팅 보낼떄 마다 sql로 확인 하는데
            // 한 번만 확인하고 그이후로는 확인했다는 데이터를 저장하는 방식으로 바꾸는게 좋아보임
            if(MysqlManager.Instance().isDMChannel(
                    Integer.parseInt(SocketConnection.Instance().userMap.get(channel).getChannelId()),
                    false)){

                String channelId = SocketConnection.Instance().userMap.get(channel).getChannelId();
                int userId = Integer.parseInt(SocketConnection.Instance().userMap.get(channel).getUserId());

                String tableName = "dm_" + channelId;
                MysqlManager.Instance().addChatData(tableName, userId, jsonObject.getString(SocketEventListener.eKey.MESSAGE.toString()), datetime);
                for (SocketChannel socket: SocketConnection.Instance().channelMap.get(channelId)) {
                    //equal메소드 작동이 잘안될수도있음 관련 에러 발생시 이거체크 braodcast쪽에서는 ==를씀
                    jsonObject.put(SocketEventListener.eKey.IS_SELF.toString(), channel.equals(socket));

                    System.out.println("channel: " +channelId+ " userId: " +userId + jsonObject.toString());
                    SocketConnection.trySendMessage(socket, jsonObject.toString());
                }
            }
        }
    }
}
